Add style checks and the minimal plugins UI

juntyr opened the pull request:
- Adds style checks to CI with black, isort, and flake8
- Fixes any style issues

- Implements plugin loading into the playground

TODOs:
- [x] fix async install on WASM
- [x] reload the "loaded plugins" list
- [x] reload read plugins
- [x] reload write plugins
- [x] reload plot plugins
- [x] reload global plugins
- [x] reload callback plugins

Future work:
- [ ] also support uploading plugins
- [ ] add list of suggested plugins
- [ ] give hints for loading plots from not-yet-installed plugins
- [ ] add custom URLs to install plugins upon page-load in the WASM version

Aggrathon commented:
> @Aggrathon Should plugin installs outside WASM be supported at all, since we cannot be sure that the user and host are the same person? Here it might be better to just error and say to use pip install or the WASM version instead.

I agree, installing plugins from the UI should be a WASM only thing

Aggrathon commented on xiplot/plugin.py:
```diff
@@ -139,6 +139,68 @@ def get_plugin_filepaths(dir_path=""):
+        # Needs to run in the background and report back via hidden UI
+        asyncio.get_running_loop().run_until_complete(
+            micropip.install(f"emfs:{plugin_path}")
+        )
```
I have tried to get this approach to work, but did not manage to find a way (that is why `scikit-learn` is lazily loaded using javascript and `setTimeout` in the WASM branch)

juntyr replied:
Yeah, I have an idea for solving it, similar to how the clustering are done. In this case, launch async install with a callback that interacts with a hidden input in the page and tells us about the success or failure which then again triggers a callback for the UI. Ugly, but it should work

juntyr commented:
> > @Aggrathon How should plugins be handled with the downloads? Do we enforce that all plugins must be uploaded as .whl files and then provide a new download option with them included and install them (with a warning emitted for each) on data upload? If we don't enforce it, how do we get the wheels for any pre-installed packages / packages installed from PyPi / from a URL?
> 
> I think we shouldn't add the plugins to the downloads. It is up to the user to remember to install all necessary plugins (although we should add error popups for missing plot types etc.). And if the user forgets, then they can just reupload the saved state after installing the necessary plugins.

Ok, we can do that. In that case, it might be nice to still remember the list of loaded plot plugins so that if there is an error loading a plot we can point to the plugin that is missing.

Aggrathon commented on xiplot/plugin.py:
```diff
@@ -88,7 +91,8 @@ def placeholder_figure(text: str) -> Dict[str, Any]:
 def get_plugins_cached(
```
Caching the lists of plugins is a premature optimisation (unless we need it for keeping track of existing callbacks). So feel free to redo this function if necessary.

juntyr commented:
@Aggrathon @TanakaAkihiro I've now finished a minimal implementation that I'd like to land so that the remaining work can be distributed.

What works:
- plugins can be installed inside WASM
- installation has a progress, success, and failure notification
- installation can happen from bundled plugins, PyPi, and .whl URLs
- the UI is updated for new plugins:
  - lists all installed plugins
  - makes all plugins (read, write, global, callback, plot) available
  - plugin overwrites are simply not executed
- there *some* is protection against concurrent and double installs
- everything is disabled in the non-WASM version, here only installed plugins are listed

What is missing but should be implemented in follow-up PRs (and probably not by me since I won't have the time):
- uploading plugins: this should be easy since it could use most of the code from uploading datasets. One problem will be the global callback for `dash_uploader`. The uploaded plugins could be put into the plugins folder and then a message could pop up to suggest selecting them in the dropdown to install them next. This could also be automated if you'd like to do that.
- listing suggested plugins: right now 'suggested' plugins are bundled. This could be changed to some metadata file that simply lists the plugin names and their installation paths. This list should be parsed into the dropdown options, and once an item has been installed successfully, it should be removed from the list. Once we have ported some plots and nice extra libraries to their separate PyPi packages, bundling *could* be removed. However, it could still be nice for development purposes.
- warning about uninstalled plugin plots: right now loading a dataset with unknown plot types will error. The download format could store wether a plot came from a plugin and if so give a warning instead that suggests installing the plugin

Since all of these are quality-of-life improvements but don't include any required functionality, I'd propose merging this PR and #34.

Aggrathon commented:
- I don't think the plugin tab should be visible in the non-WASM version. It takes too much space for something that is not working. I understand the "advertising" value, but since it is a new top-level tab (instead of e.g. being part of the settings tab) it is too visible.
- On narrower window sizes the plugin instructions get cut off. Maybe add a hovertext or something.
- The list of installed plugins is functional, but ugly. Currently an item is `[hook_type] entry-point-key: entry-point-target`. A simple improvement would be to remove either the `entry-point-key` or the `entry-point-target`. But ideally I would suggest only showing the _package name / .whl name / .whl url_.
- If uploading is not working, just hide the box.
- I tried installing a package that is not a plugin, but adds functionality, namely `fastparquet`. And I'm happy to report that the UI is correctly updated and the loading of .parquet files works as expected. However, it is not added to the list of plugins and trying to save .parquet files raises an IOError (complains about "file closed" for some reason?), but this is maybe for a future bugfix.
- Suggestion for future work: What if we could construct links like https://edahelsinki.fi/xiplot?plugin=package_a&plugin=package_b that would automatically load the specified plugins?

Aggrathon commented on xiplot/plugin.py:
```diff
@@ -88,20 +95,23 @@ def placeholder_figure(text: str) -> Dict[str, Any]:
 def get_plugins_cached(
```
The idea with the `plugins.py` file was to gather everything a plugin-implementer would need in a single place. Now this file mostly contains internal xiplot functions for loading plugins. Could those be moved? (e.g. to `tabs/plugins.py`)

Aggrathon replied:
I don't have anything other to add at the moment, the functionality of the PR looks great!

juntyr commented:
> * I don't think the plugin tab should be visible in the non-WASM version. It takes too much space for something that is not working. I understand the "advertising" value, but since it is a new top-level tab (instead of e.g. being part of the settings tab) it is too visible.

That is a possibility and would simplify some code, though it would still be nice to have a list of loaded plugins in the non-WASM version.

>     * If uploading is not working, just hide the box.

If we hide the plugins in non-WASM, this issue is resolved in my opinion. We should provide uploads in WASM unconditionally, but not hide them right now because it serves as an implementation reminder.

>     * On narrower window sizes the plugin instructions get cut off. Maybe add a hovertext or something.

Sounds good to me.

>     * The list of installed plugins is functional, but ugly. Currently an item is `[hook_type] entry-point-key: entry-point-target`. A simple improvement would be to remove either the `entry-point-key` or the `entry-point-target`. But ideally I would suggest only showing the _package name / .whl name / .whl url_.
>     * ... However, it is not added to the list of plugins ...

This is a fundamental limitation of our plugin discovery at the moment. We only know about plugins that provide us with the known xiplot hooks. While we could list every installed PyPi plugin, I don't think this would be useful. What could be done is to list all entry point targets, derive the package names from that, and then list `package name: (hook-type),*`

>     * I tried installing a package that is not a plugin, but adds functionality, namely `fastparquet`. And I'm happy to report that the UI is correctly updated and the loading of .parquet files works as expected. However, it is not added to the list of plugins and trying to save .parquet files raises an IOError (complains about "file closed" for some reason?), but this is maybe for a future bugfix.

I don't like that this is effectively a silent plugin at the moment, i.e. installing it has an effect but we cannot detect really detect it, unless we special-case the plugin discovery to check for parquet support. I'd rather us publish a wrapper plugin that just imports the dependency. The bug fix is out of scope right now.

>     * Suggestion for future work: What if we could construct links like https://edahelsinki.fi/xiplot?plugin=package_a&plugin=package_b that would automatically load the specified plugins?

Sounds good to me, out of scope for now.

Aggrathon commented:
> That is a possibility and would simplify some code, though it would still be nice to have a list of loaded plugins in the non-WASM version.

A list of installed plugins in the non-WASM version is a nice touch. However, personally I don't think it is that valuable, especially if removing it would lead to improved UI and, possibly, a better list for the WASM version.

Aggrathon commented:
> The bug is in fastparquet, but I added a workaround for it. Essentially, I subclass `BytesIO` to disallow it from being closed while the df is written to it. Afterwards, I manually do the cleanup with a context helper class. It seems to work.

Oh, that ended up being more convoluted than expected (and in hindsight maybe not that necessary to work around). But the solution you found is pretty cool!

Aggrathon commented:
> I've now moved the list to the settings menu if the plugins tab is not loaded because loading new plugins is unsupported. The problem with a better plugin list, which includes installed packages that don't register any hooks, is that we'd have to diff the micropip list after an install from before the launch to get all packages. This would include the lazy sklearn load, and any dependencies any plugin might have. I think it would be far too cluttered too quickly.

Leaving improvements to the plugin list for future work is okay, but it already looks pretty great (one nitpick would be to parse the path with `path.split(":")[0].split(".")[0]` otherwise the [Slisemap plugin](https://github.com/edahelsinki/slisemap_interactive) shows up as `slisemap_interactive.xiplot`). 

What I would have done is have a dumb list that gets appended to whenever the user clicks "install" (after a successful install just add whatever was in the input, therefore ignoring dependencies), but this would not work for the non-WASM list.

Aggrathon commented on xiplot/tabs/settings.py:
```diff
@@ -75,43 +78,64 @@ def register_callbacks(app: Dash, df_from_store, df_to_store):
+                                clearable=False,
+                                searchable=False,
+                                placeholder=(
+                                    "Check the list of already installed"
+                                    " plugins"
```
```suggestion
                                    "Check the list of installed plugins"
```

Aggrathon commented on xiplot/tabs/settings.py:
```diff
@@ -121,7 +145,38 @@ def create_layout():
+        name,
+        path,
+    ) in get_all_loaded_plugins():
+        plugins[path.split(":")[0]].add(kind)
```
```suggestion
        plugins[path.split(":")[0].split(".")[0]].add(kind)
```

juntyr commented:
@Aggrathon @TanakaAkihiro This PR is now ready for final review and to be merged.

TanakaAkihiro reviewed (approved):
LGTM!

Aggrathon commented on xiplot/tabs/plugins.py:
```diff
@@ -0,0 +1,406 @@
+            "Loading new plugins is only supported in WASM"
+        )
+
+    if len(plugin_source.split()) != 1:
+        raise ValueError("Plugin source must be a URL or PyPi package name")
+
+    if (
+        Path(plugin_source).exists()
+        and Path(plugin_source).name != plugin_source
+    ):
+        raise ValueError("Plugin source must be a URL or PyPi package name")
```
Is there a reason for not letting micropip handle the parsing of the string (and throw a more detailed error)?

Aggrathon commented:
I tried having two "recommended" plugins in the WASM branch, but as soon as I loaded one of them the list of recommended plugins was cleared. This is not a show stopping bug (since we don't have multiple recommended plugins) but it should at least be mentioned for future work.

juntyr commented:
> I tried having two "recommended" plugins in the WASM branch, but as soon as I loaded one of them the list of recommended plugins was cleared. This is not a show stopping bug (since we don't have multiple recommended plugins) but it should at least be mentioned for future work.

That is a bug. In the process I discovered another issue and fixed both. Thanks for catching it! I also renamed the `dir_path` variable to `data_dir`.

Aggrathon commented:
I've found a major issue. Any dynamically added callback is not registered on the client side. Meaning any plugin type requiring callbacks, such as plots, do not work. Below is a simple modification to the test plugin that should demonstrate this:

```python
import pandas as pd
from dash import html, Input, Output, MATCH


def plugin_load():
    def read(data):
        print("Reading test data")
        return pd.DataFrame({"x": [1, 2, 3]})

    return read, ".test"


def plugin_write():
    def write(df, file):
        print("Writing test data")

    return write, ".test", "example/none"


def create_global():
    return html.Div("TEST PLUGIN", style={"display": "none"}, id="test-plugin-global")


def register_callbacks(app, df_from_store, df_to_store):
    print("The test-plugin has registered a callback")


class Plot:
    @classmethod
    def name(cls) -> str:
        return "  TEST PLUGIN"

    @staticmethod
    def register_callbacks(app, df_from_store, df_to_store):
        @app.callback(
            Output({"index": MATCH, "type": "test_counter"}, "children"),
            Input({"index": MATCH, "type": "test_button"}, "n_clicks"),
        )
        def counter(num):
            return f"{num} clicks"

        print("The test-plugin plot has registered a callback")

    @staticmethod
    def create_new_layout(index, df, columns, config=dict()):
        return html.Div(
            [
                html.Button(["Test Plot"], id={"index": index, "type": "test_button"}),
                html.Span(["No clicks"], id={"index": index, "type": "test_counter"}),
            ]
        )
```

juntyr commented:
@Aggrathon Well dang. I tried for a while to get it to work, but it doesn't (the problem is that the frontend would need to refetch the "/_dash-dependencies" endpoint to get the new callbacks - I have no idea how to force it to do that though). I have figured out a way to do the page reload though. So my suggestion is to remove all dynamic plugin update code, to track which plugins have been loaded since the last reload and are thus not accessible yet, and to give a big reload button for them. We should probably also disallow the read and write plugins from working immediately, just for consistency.

juntyr commented:
I'll implement this next week.

juntyr commented:
@Aggrathon I've now implemented the manual reload variant and stripped out the code for the dynamic plugin loads. I hope that this version is now ready to be merged.

Aggrathon commented on xiplot/tabs/plugins.py:
```diff
@@ -0,0 +1,477 @@
+                                "Reload χiplot",
+                                id="plugin-reload-button",
+                                n_clicks=0,
+                                className="button",
+                                disabled=True,
+                            ),
+                            dmc.Modal(
+                                title="Do you really want to reload χiplot?",
+                                id="plugin-reload-modal",
+                                zIndex=10000,
+                                children=[
+                                    dmc.Text(
+                                        "Please ensure that you have"
+                                        " downloaded your plots and data so"
+                                        " that nothing is lost. Any plugins"
+                                        " that you have installed will become"
+                                        " loaded after the reload."
+                                    ),
+                                    dmc.Space(h=20),
+                                    dmc.Group(
+                                        [
+                                            dmc.Button(
+                                                "Reload",
+                                                id="plugin-reload-accept",
+                                            ),
+                                            dmc.Button(
+                                                "Cancel",
+                                                color="red",
+                                                variant="outline",
+                                                id="plugin-reload-cancel",
+                                            ),
+                                        ],
+                                        position="right",
+                                    ),
+                                ],
+                            ),
+                        ),
+                        title="Loaded Plugins",
+                        css_class="dash-dropdown",
+                    ),
```
After installing a plugin I immediately opened the "loaded plugins list" and it wasn't there. It only appeared after a reload. This does make sense, but I think calling the list "active plugins" instead of "loaded plugins" would help with this confusion.

```suggestion
                    layout_wrapper(
                        component=FlexRow(
                            dcc.Dropdown(
                                get_loaded_plugin_options(),
                                id="loaded_plugins",
                                clearable=False,
                                searchable=False,
                                placeholder=(
                                    "Check the list of activated plugins"
                                ),
                            ),
                            html.Button(
                                "Reload χiplot",
                                id="plugin-reload-button",
                                n_clicks=0,
                                className="button",
                                disabled=True,
                            ),
                            dmc.Modal(
                                title="Do you really want to reload χiplot?",
                                id="plugin-reload-modal",
                                zIndex=10000,
                                children=[
                                    dmc.Text(
                                        "Please ensure that you have"
                                        " downloaded your plots and data so"
                                        " that nothing is lost. Any plugins"
                                        " that you have installed will become"
                                        " activated after the reload."
                                    ),
                                    dmc.Space(h=20),
                                    dmc.Group(
                                        [
                                            dmc.Button(
                                                "Reload",
                                                id="plugin-reload-accept",
                                            ),
                                            dmc.Button(
                                                "Cancel",
                                                color="red",
                                                variant="outline",
                                                id="plugin-reload-cancel",
                                            ),
                                        ],
                                        position="right",
                                    ),
                                ],
                            ),
                        ),
                        title="Active Plugins",
                        css_class="dash-dropdown",
                    ),
```

Aggrathon commented on xiplot/tabs/plugins.py:
```diff
@@ -0,0 +1,477 @@
+                            color="yellow",
+                            title="Information",
+                            message=(
+                                "You can now load the newly installed plugin"
```
```suggestion
                                "You can now activate the newly installed plugin"
```

Aggrathon commented on xiplot/tabs/plugins.py:
```diff
@@ -0,0 +1,477 @@
+                                id="plugin-reload-modal",
+                                zIndex=10000,
+                                children=[
+                                    dmc.Text(
+                                        "Please ensure that you have"
+                                        " downloaded your plots and data so"
+                                        " that nothing is lost. Any plugins"
+                                        " that you have installed will become"
+                                        " loaded after the reload."
+                                    ),
```
A list of the packages that would become activated would be nice (here or somewhere else). But that could also be left for future work.

juntyr replied:
Good idea - I'll work on this tomorrow and then merge the PR

Aggrathon reviewed (approved):
LGTM! (even without my latest suggestions)
